lib/class-core.php

```diff
@@ -0,0 +1,223 @@
+			if ( empty( $current_site['term_id'] ) ) {
+				return false;
+			}
+
+			// If we have the data we need in cache, return it right away.
+			if ( ! empty( $current_site[ $property ] ) ) {
+				return $current_site[ $property ];
+			}
+
+			// Get the site term for the current site.
+			$term = get_term( $current_site['term_id'], Site::instance()->name );
+
+			// Return false on error.
+			if ( is_wp_error( $term ) || empty( $term->slug ) ) {
+				$this->site_term = false;
+			} else {
+				$this->site_term = $term;
+			}
+		}
+
+		// If we want a specific property, return that.
+		if ( $this->site_term && $property ) {
+			return isset( $this->site_term->$property ) ? $this->site_term->$property : null;
+		} else {
+			return $this->site_term;
+		}
+	}
+
+	/**
+	 * Get the current site from the cached option.
+	 *
+	 * @return array {
+	 *     Basic information about the current site.
+	 *
+	 *     @type int $term_id The site's term id.
+	 *     @type string $name The site's term name (same as the current host).
+	 *     @type string $slug The site's term slug.
+	 * }
+	 */
+	public function get_current_site_from_cache() {
+		$domains = get_option( 'split_domain_sites', [] );
+		if ( empty( $domains ) ) {
+			return;
+		}
+
+		$current_host = parse_url( home_url(), PHP_URL_HOST );
+		if ( ! empty( $domains[ $current_host ] ) ) {
+			$return = $domains[ $current_host ];
+			$return['name'] = $current_host;
+			return $return;
+		} else {
+			return false;
+		}
+	}
+
+	/**
+	 * Get the 'site-domain' WP_Term for a given (or the current) post.
+	 *
+	 * @param  \WP_Post|int $post Post object or ID.
+	 * @return \WP_Term|false Term object on success, false on failure.
+	 */
+	public function get_post_site( $post = null ) {
+		$post = get_post( $post );
+		$sites = get_the_terms( $post, Site::instance()->name );
+		if ( is_wp_error( $sites ) || empty( $sites ) ) {
+			return false;
+		} else {
+			return reset( $sites );
+		}
+	}
+
+	/**
+	 * Get the default site term.
+	 *
+	 * @return \WP_Term|false Term object on success, false on failure.
+	 */
+	public function get_default_site() {
+		$default_term_id = Settings::instance()->get_setting( 'default' );
+		if ( is_int( $default_term_id ) ) {
+			$term = get_term( $default_term_id, Site::instance()->name );
+			if ( $term && ! is_wp_error( $term ) ) {
+				return $term;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Redirect to the canonical version of the current url if the site is not
+	 * correct.
+	 */
+	public function redirects() {
+		if ( is_singular() ) {
+			$post_site = $this->get_post_site();
+			if ( ! $post_site ) {
+				$post_site = $this->get_default_site();
+				if ( ! $post_site ) {
+					return;
+				}
+			}
+
+			$current_site = $this->get_current_site_term();
+			if ( ! $current_site || $post_site->name === $current_site->name ) {
+				return;
+			}
+
+			wp_redirect( str_replace( $current_site->name, $post_site->name, get_permalink() ), 301 );
```

Comment:
You consider using `wp_safe_redirect` so you could pull from a list of approved hosts via the plugin? Just a thought, VIP might like that.